api_app/services/aad_authentication.py

```diff
@@ -1,60 +1,75 @@
 import base64
 import logging
+from typing import List
 import jwt
 import requests
 import rsa
 
 from fastapi import Request, HTTPException, status
-from fastapi.security import OAuth2AuthorizationCodeBearer
+from msal import ConfidentialClientApplication
 
+from services.access_service import AccessService, AuthConfigValidationError
 from core import config
 from db.errors import EntityDoesNotExist
-from models.domain.authentication import User
+from models.domain.authentication import User, RoleAssignment
+from models.domain.workspace import Workspace, WorkspaceRole
 from resources import strings
 from api.dependencies.database import get_db_client_from_request
 from db.repositories.workspaces import WorkspaceRepository
 
 
-class AzureADAuthorization(OAuth2AuthorizationCodeBearer):
+class AzureADAuthorization(AccessService):
     _jwt_keys: dict = {}
 
-    _default_app_reg_id = None
+    require_one_of_roles = None
 
-    def __init__(self, aad_instance: str, aad_tenant: str, auto_error: bool = True, app_reg_id: str = None):
+    def __init__(self, auto_error: bool = True, require_one_of_roles: list = None):
         super(AzureADAuthorization, self).__init__(
-            authorizationUrl=f"{aad_instance}/{aad_tenant}/oauth2/v2.0/authorize",
-            tokenUrl=f"{aad_instance}/{aad_tenant}/oauth2/v2.0/token",
-            refreshUrl=f"{aad_instance}/{aad_tenant}/oauth2/v2.0/token",
-            scheme_name="oauth2", scopes={
-                f"api://{config.API_CLIENT_ID}/Workspace.Read": "List and Get TRE Workspaces",
-                f"api://{config.API_CLIENT_ID}/Workspace.Write": "Modify TRE Workspaces"
-            },
+            authorizationUrl=f"{config.AAD_INSTANCE}/{config.AAD_TENANT_ID}/oauth2/v2.0/authorize",
+            tokenUrl=f"{config.AAD_INSTANCE}/{config.AAD_TENANT_ID}/oauth2/v2.0/token",
+            refreshUrl=f"{config.AAD_INSTANCE}/{config.AAD_TENANT_ID}/oauth2/v2.0/token",
+            scheme_name="oauth2",
```

Comment:
Should revisit scopes - do we need to verify scopes? The scopes need to be different if its a workspace operation vs TRE API operation.

Reply:
Lets look at integrating scoeps and moving to https://intility.github.io/fastapi-azure-auth/ as a seperate issue.

Reply:
https://github.com/microsoft/AzureTRE/issues/1074